javascript/ql/src/semmle/javascript/security/dataflow/Xss.qll

```diff
@@ -0,0 +1,225 @@
+/**
+ * Provides classes and predicates used by the XSS queries.
+ */
+
+import javascript
+
+/** Provides classes and predicates shared between the XSS queries. */
+module Shared {
+  /** A data flow source for XSS vulnerabilities. */
+  abstract class Source extends DataFlow::Node { }
+
+  /** A data flow sink for XSS vulnerabilities. */
+  abstract class Sink extends DataFlow::Node {
+    /**
+     * Gets the kind of vulnerability to report in the alert message.
+     *
+     * Defaults to `Cross-site scripting`, but may be overriden for sinks
+     * that do not allow script injection, but injection of other undesirable HTML elements.
+     */
+    string getVulnerabilityKind() { result = "Cross-site scripting" }
+  }
+
+  /** A sanitizer for XSS vulnerabilities. */
+  abstract class Sanitizer extends DataFlow::Node { }
+}
+
+/** Provides classes and predicates for the DOM-based XSS query. */
```

Comment:
I assume the three modules below are mostly cut-pasted, and have not looked into their details.

Reply:
Correct.